replace_with/src/lib.rs

```diff
@@ -0,0 +1,11 @@
+use std::ptr;
+use std::fmt;
+
+pub fn replace_with<T: fmt::Debug, F: FnOnce(T) -> T>(t : &mut T, f: F) {
+    let t_ptr = t as *mut T;
+    unsafe { 
+        let new_int = ptr::read(t_ptr);
+        ptr::write(t_ptr, f(new_int));
+    }
```

Comment:
Nice!
Ross discovered that you can use `t` in place of `t_ptr` as the argument to `write` and `read`. Apparently an `&mut` can coerce to a `*const` or `*mut` and function call sites trigger coercion.